### PR TITLE
Directory layout change, add composer config and more

### PR DESCRIPTION
**Pull request by jrfnl**
### Move source files to src subdirectory

This will make it more straight-forward to set up autoloading with Composer and as the package is not published as a "Package manager ready" package prior to this, IMO this directory layout change should not be seen as a breaking change.

### Move test file to tests subdirectory

This allows for all test related files to be excluded from git archives in one go and will also allow (in a potential future) for additional test scripts/classes to be added and test autoloading via Composer to be set up.

### Add composer.json configuration file

A `composer.json` file defines the project specifications for the currently most commonly used package manager in the PHP world.

Aside from the obvious:
* I've set the license to the [Creative Commons Public Domain Dedication and Certification](https://spdx.org/licenses/CC-PDDC.html) license, which is the closest I could find to an official "public domain" license.
* I've added an autoload section to allow dependents on the package to use the Composer generated autoload file to load the `PassworfHash` class (instead of having to manually `require` the file).
* I've added a script to make it self-evident to potential contributors how to run the tests.

Running `composer install` will generate a `composer.lock` file and create the autoload files in a `vendor` directory.
If the project would have dependencies itself, those would also be installed in the `vendor` directory.

* As a general best-practice it is recommended **not** to commit the `vendor` directory to git.
* Whether or not to commit the `composer.lock` file is highly dependent on the package type. For a generic library package which supports a wide range of PHP versions, I would recommend not to commit the `composer.lock` file.

With the above in mind, I'm also adding a `.gitignore` file to prevent these files from being accidentally committed to the repo.

### Add a .gitattributes file

GitHub/Git allows for creating package archives as zip files. Those archives are also used by Packagist.

It is generally considered good practice for those package to only contain source files.
Tests, CI related files and tooling configuration files should be excluded from those archives to keep them lean.

This `.gitattributes` files does exactly that and prevents files which are not needed by dependencies from being included in the archives.

As a bonus, line ending normalization has been added as well.

### CODE REVIEW

File: composer.json
Comment (solardiz) [inline composer.json:3]: Delete trailing whitespace after "5.5" and before the closing quote.

File: tests/test.php
Comment (solardiz) [inline tests/test.php:9]: Instead of the double use of `dirname`, do `dirname(__FILE__) . '../src/PasswordHash.php'`

**Review by solardiz — CHANGES_REQUESTED** [review]
Thank you!

I've made some comments - please make the changes accordingly.

Also, please wrap commit description lines at 75, so that `git log` can be read without needing horizontal scrolling on an 80-wide terminal. I didn't nitpick like this on your previous commits, but now that you continue to contribute I think we should make further commits cleaner.

File: composer.json
Comment (jrfnl) [reply composer.json:13]: Thanks for catching that.

File: .gitattributes
Comment (jrfnl) [reply .gitattributes:12]: The `.gitattributes` file is unrelated to GitHub Actions. It's about keeping the development related files outside of other people's production environment (where they don't belong and could potentially even create a security risk, as has happened before).

Please read the articles which I linked to in the documentation of the file.
* https://www.reddit.com/r/PHP/comments/2jzp6k/i_dont_need_your_tests_in_my_production
* https://blog.madewithlove.be/post/gitattributes/

Comment (solardiz) [reply .gitattributes:12]: Sure I know what `.gitattributes` is, and we use it in some other repos here. Good point about test files being a security risk for PHP projects in particular, where the files could commonly end up directly accessible via a web server. OK, let's do it the way you propose. However, then there's a new decision to make: should future phpass official tarballs be made with `git archive` (which I thought was the intent here) or maybe not (so they retain what they used to have, with my expectation being that only the one needed source file would be taken out and incorporated in a project's tree)?

**Comment by solardiz** [comment]
> who still uses an 80-wide terminal in 2023 ?

Well, I usually set mine to 132 these days, but that doesn't change a thing for you - your lines were way longer than 132. So perhaps your question is better worded as "who uses `git log` in a terminal these days", to which I'd reply that I and a lot of other people do.

My current approach is to allow longer than 80 (but shorter than 132) source code lines, but keep documentation and even source code comments to within 80.

Keeping git commit messages to within 75 is also a convention used by the Linux kernel, where git originated. (`git log` indents them by 4 spaces, so this brings us to 79.)

It sure is a problem if that would discourage you from contributing. Maybe I should just accept that phpass would become more of your project than mine from this point on, and then you do whatever in terms of conventions.

File: .gitattributes
Comment (solardiz) [reply .gitattributes:12]: Well, rather than GPG-sign an archive downloaded from GitHub, I'd review my local copy and use `git archive` and sign that. Anyway, I could just be tarring up the whole tree excluding `.git` for such purpose, and people who don't care to have signed content can use the GitHub-generated archives that would also exclude tests.

**Review by solardiz — APPROVED** [review]
I think this is ready to merge. I do notice some typos in the commit messages (like "be be"), but I don't care that much.

**Comment by jrfnl** [comment]
> I think this is ready to merge. I do notice some typos in the commit messages (like "be be"), but I don't care that much.

Fixed that one ;-) And yes, I'm dyslectic, so typos and such happen to me, no matter how often I re-read commit messages... 🤷🏻‍♀️

**Comment by jrfnl** [comment]
> It sure is a problem if that would discourage you from contributing. Maybe I should just accept that phpass would become more of your project than mine from this point on, and then you do whatever in terms of conventions.

I'm not looking to become a maintainer of _yet another_ package, though I sure would be happy to contribute and collaborate if we can figure out a way to make this point workable for both.

**Comment by solardiz** [comment]
> make this point workable for both.

Well, I expressed my preference and convention so far, but I don't insist on it. I'll leave it up to you what to do about it going forward.

Is phpass ready for a new release now, perhaps 0.6 or 1.0? Should we push a tag?

**Comment by solardiz** [comment]
Yes, I was thinking to release a 0.5.2 with just the #2 fix as well, so that we're not forcing all of the other changes onto someone merely needing the deprecation warning fix.

> I'd say 0.6 for now and wait with the 1.0 until there is a more comprehensive test suite ?

Are we planning to have a more comprehensive test suite? What would it add?

**Comment by jrfnl** [comment]
> Are we planning to have a more comprehensive test suite? What would it add?

Not sure about the "we" or the "planning", but it is a suggestion I made in a [previous comment](https://github.com/openwall/phpass/pull/2#issuecomment-1476526363) and I would be happy to contribute to it.

> What with the huge amount of changes, especially deprecations, over the last few years in PHP, I can imagine, it might also be interesting/worth exploring to expand the test suite so it can function as an early warning system for new issues which crop up in new PHP versions. Would you be open to that ?

**Comment by costdev** [comment]
Adding a more comprehensive test suite would be good to report which tests failed, why, and serve as an early-warning system as new PHP versions are released. Issues would then be caught in this repo, and could be fixed prior to the PHP version's release.

It would also make it easier to contribute to anything that needs to be updated in future, and also make writing tests for a change more familiar to potential contributors.

Writing effective tests means getting a list of different possible inputs/outputs to test different scenarios. This is where your knowledge would be critical @solardiz. I know that there are some [test vectors](https://www.schneier.com/wp-content/uploads/2015/12/vectors-2.txt) available - although I don't have enough knowledge on this to determine what the inputs and expected outputs are from that. There may also be scenarios where the length of an input is incorrect and such - but you would have a better idea about that. 🙂

**Comment by solardiz** [comment]
@jrfnl Sure let's enable detection of the deprecation notices in GitHub Actions.

@costdev Those test vectors you link to are for Blowfish, not for bcrypt, and are not usable here. We can reasonably add a test with an explicit bcrypt hash to `test.php` here - right now, we end up testing a bcrypt hash implicitly when run on systems that support it in PHP, but we do not against a static test vector. Adding such a test, if it's mandatory, would make tests fail on versions of PHP too old to support bcrypt (unless run on an OS that provides bcrypt to PHP). This would only make a difference for PHP 5.0 (we don't support older) to 5.2.x (since 5.3+ has bcrypt of its own).

The bcrypt code in PHP proper, which is based on code I wrote with minor changes, includes a runtime self-test against a static test vector (which I added to there). So when we're running on recent PHP, we already implicitly perform a test of bcrypt against a static test vector.

We might want to add a similar runtime self-test for the fallback hashes in phpass as well, although this would normally only matter on PHP below 5.3. When testing of phpass portable hashes against their stored values on modern PHP, which is probably the most relevant remaining use case for this package, the worst that can realistically happen on miscomputation of those hashes is failed authentication. It is unlikely that a bug would mysteriously compute the right hash on a wrong password - well, maybe except for cases when the passed wrong password includes in it the correct password plus garbage.

I am just thinking out loud here and sharing. I want to keep the actual runtime code simple, so adding more runtime self-tests would need to be justified. I understand you two were not referring to _runtime_ tests, but that's what I care about more.

**Comment by solardiz** [comment]
> Yes, I was thinking to release a 0.5.2 with just the #2 fix as well

In hindsight, we should have updated the source code comment to say 0.5.2 in that commit, or at least before adding more commits. Now we'd need a branch for that, or release 0.5.2 that doesn't exactly match any git commit (would differ in that comment only). Well, or we could rewrite history, assuming that very few people pulled this repo in these two days.

**Comment by jrfnl** [comment]
> > Yes, I was thinking to release a 0.5.2 with just the #2 fix as well
> 
> In hindsight, we should have updated the source code comment to say 0.5.2 in that commit, or at least before adding more commits. Now we'd need a branch for that, or release 0.5.2 that doesn't exactly match any git commit (would differ in that comment only). Well, or we could rewrite history, assuming that very few people pulled this repo in these two days.

Considering there are only [7 forks and only one of those is up to date](https://github.com/openwall/phpass/forks?activity_threshold=5y&include=active&page=1&sort_by=stargazer_counts) (mine), I'd say: go for the history rewrite and I'll sync my fork.

**Comment by costdev** [comment]
Thanks for the clarification @solardiz - Much appreciated!

**Pull request merged**
